### ISSUE TITLE
Sparse array and sparse solver in AD

### ISSUE THREAD

**Issue by lonelykid** [kailaix/ADCME.jl#25]
Hi, do you know any AD tool supporting sparse array? I just did a simple demo for FEM with Google/JAX, which has the same backend as Tensorflow. The problem is, without sparse array for global stiffness, dense solvers are so slow that only toy problem can be solved. A linear problem with 10k elements takes almost two minutes to solve, while it takes less than 1 second with numpy.

Tensorflow does support sparse tensor, but its linear solver that not support it. There is an issue saying they are working on it, though.

**Comment by kailaix** [comment]
Hi @lonelykid just saw the issue. Did you solve the problem?
Actually, if you use `SparseTensor` to represent sparse matrices in ADCME, you automatically invoke sparse linear solvers (SparseLU for square matrices and SparseQR for least-square problems). Also, the gradients were implemented so you can freely backpropagate gradients through the right hand side or the coefficient matrix nonzero values. I added a documentation note here [https://kailaix.github.io/ADCME.jl/dev/sparse/#Sparse-Solvers-1](https://kailaix.github.io/ADCME.jl/dev/sparse/#Sparse-Solvers-1).

**Comment by lonelykid** [comment]
Nice job! I found this is not yet implemented in tensorflow yet. I found something in HIPS/autograd and is about to try. They didn't implement but it's not hard to add support for sparse solvers. 
Another question, when using a dense array, most of the time is spent on assembling because of poor performance of lax.scan(or tf.while_loop). If we use sparse array, does this mean we don't need to worry about assembling at all since we could just use repeated indices and supposedly the solver will take care of it? It seems you are working a custom operator of assembling.

**Comment by kailaix** [comment]
Yes, you do not need to worry about repeated indices. That's being cared for by `SparseTensor`.
As for custom operator of assembling, I guess you mean [`SparseAssembler`](https://kailaix.github.io/ADCME.jl/dev/api/#ADCME.SparseAssembler-Tuple{}). Sometimes each element may contribute to a different number of nonzeros in the global stiffness matrices (e.g., on the boundary, or different quadrature rules), thus it is inefficient to use `TensorArray` to store all the local stiffness matrices and then aggregate. Instead, you can call `accumulator` (one output of `SparseAssembler`) in your `while_loop` and finally call `creator` to get a sparse matrix. Unfortunately, this function currently does not back-propagate gradients (in fact, it is intended for assembling sparse Jacobians instead of stiffness matrices).

**Comment by lonelykid** [comment]
I'm not sure when you need the Assembler. With sparse tensor, you are just concatenating arrays of indices and values. You mean this is not sufficient enough?

**Comment by kailaix** [comment]
I have not found a way to do that efficiently in `while_loop`. I used `TensorArray` to store indices and values, but each iteration in `while_loop` may contribute an array of different lengths  (in a use case, it varies from 4 to **24**). Then you have to create "#elements by **24**" `TensorArray`s for rows, cols, and values for assembling. You cannot simply concatenate arrays of different lengths with `while_loop`; even if you can, that's not efficient because that will create a lot of temporary variables (dynamic memory allocation). 
If you found a way to do it efficiently and would love to share, I'd appreciate very much :)

**Comment by lonelykid** [comment]
Got it. Thank you.

**Comment by lonelykid** [comment]
You are right: in order to pursue performance, it seems that you have to use fixed length. I haven't try this yet. Can you use tf.vectorized_map to parallelize the calculation of local jocobian? And then use tf.concat once. You don't need really need a while_loop.

**Comment by kailaix** [comment]
That's very interesting. It looks like `tf.vectorized_map` is a more efficient implementation of data independent parallel loops. Thanks for pointing out. 
I did not use `tf.cond`. It's too complicated to show a full script but to illustrate, I gave a relevant example here 
```
a = constant(rand(10))
b1 = sum(a[1:2]) # boundary term, centered at 1
g1 = gradients(b1, a)
b2 = sum(a[1:3]) # non-boundary term, centered at 2
g2 = gradients(b2, a)
sess = Session(); init(sess)
run(sess, g1) # length 2 vectors
run(sess, g2) # length 3 vectors
```

**Comment by lonelykid** [comment]
I see. Thanks. Basically, you hardcoded different cases separately and combine them. I was thinking about switching based on inputs, e.g. element types, quadrature rules.

**Comment by kailaix** [comment]
Yes. I will explore `tf.vectorized_map` a bit more. Thanks for the information.
